<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Methods for use with discovery GWAS • winnerscurse</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Methods for use with discovery GWAS">
<meta property="og:description" content="winnerscurse">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">winnerscurse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/amandaforde/winners_curse" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Methods for use with discovery GWAS</h1>
            
      
      
      <div class="hidden name"><code>winners_curse_methods.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://amandaforde.github.io/winnerscurse/" class="external-link">winnerscurse</a></span><span class="op">)</span></span></code></pre></div>
<p>This vignette demonstrates how our R package,
<code>winnerscurse</code>, can be used to obtain new adjusted
association estimates for each SNP when only the summary statistics from
a single genetic association study, namely the discovery GWAS, are
available. In order to do so, we create a toy data set using the
function <code>sim_stats</code>, which is included in the package, and
subsequently, illustrate how a user could engage with each of the
package’s winner’s curse correction functions using this data set.</p>
<p>The methods currently available for implementation are:</p>
<ol style="list-style-type: decimal">
<li>Conditional Likelihood methods - <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2665019/" class="external-link">Ghosh
<em>et al.</em> (2008)</a>
</li>
<li>Empirical Bayes method - <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4048064/" class="external-link">Ferguson
<em>et al.</em> (2013)</a>
</li>
<li>FDR Inverse Quantile Transformation method - <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5013908/" class="external-link">Bigdeli
<em>et al.</em> (2016)</a>
</li>
<li>Bootstrap method - <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>
</li>
</ol>
<p>An important feature which distinguishes these four methods is
detailed in the table below:</p>
<table class="table">
<colgroup>
<col width="53%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Adjusts <span class="math inline">\(\beta\)</span> estimate of
<em>only</em> SNPs with <span class="math inline">\(p\)</span>-values
less than specified threshold, <span class="math inline">\(\alpha\)</span>
</th>
<th>Adjusts <span class="math inline">\(\beta\)</span> estimate of
<em>all</em> SNPs</th>
</tr></thead>
<tbody><tr class="odd">
<td><ul>
<li>Conditional Likelihood methods</li>
</ul></td>
<td><ul>
<li>Empirical Bayes method</li>
<li>FDR Inverse Quantile Transformation method</li>
<li>Bootstrap method</li>
</ul></td>
</tr></tbody>
</table>
<div class="section level3">
<h3 id="creating-a-toy-data-set">Creating a toy data set<a class="anchor" aria-label="anchor" href="#creating-a-toy-data-set"></a>
</h3>
<ul>
<li><p>The function <code>sim_stats</code> provides users with an
opportunity to construct various simulated data sets of GWAS summary
statistics which the correction methods can then be applied to. This
allows users to perhaps gain a better understanding of each method
before applying them to their own real data sets.</p></li>
<li>
<p><code>sim_stats</code> constructs a large set of independent
SNPs, in which there exists a polygenic background of a specified
portion of these SNPs for the trait of interest while all other SNPs
demonstrate no effect. When considering obtaining a simulated set of
summary statistics for solely the discovery GWAS, <code>sim_stats</code>
requires the user to specify four arguments, namely <code>nsnp</code>,
<code>h2</code>, <code>prop_effect</code> and <code>nid</code>:</p>
<ol style="list-style-type: decimal">
<li>
<code>nsnp</code>: the total number of SNPs</li>
<li>
<code>h2</code>: the heritability of the trait, i.e. the variance
explained in the trait by all SNPs</li>
<li>
<code>prop_effect</code>: the proportion of the total number of SNPs
truly associated with the trait</li>
<li>
<code>nid</code>: the number of samples in the study</li>
</ol>
</li>
<li><p>With the above specifications, a set of rather realistic summary
statistics can be obtained as shown below. For convenience purposes, the
following default values have been supplied for the function’s
arguments: <code>nsnp=10^6</code>, <code>h2=0.4</code>,
<code>prop_effect=0.01</code>, <code>nid=50000</code>. These values
simulate summary statistics for 1,000,000 independent SNPs in which 1%
have a true effect on this trait with moderate heritability and a total
of 50,000 individuals have been sampled in the study.</p></li>
<li><p><strong>Note:</strong> For further information regarding
simulating GWAS summary statistics, refer to the <a href="https://doi.org/10.1371/journal.pgen.1010546.s039" class="external-link">text
supplement</a> of <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>. This function has been designed based on the
details described under the heading <em>‘Quantitative trait with
independence assumed’</em>.</p></li>
<li><p><code>sim_stats</code> outputs a list with three components. When
only the discovery GWAS is considered, the third element in this list is
merely <code>NULL</code>, and so we are just concerned with the first
two, <code>true</code> and <code>disc</code>. <code>true</code> is a
data frame with two columns, <code>rsid</code> which contains
identification numbers for each SNP and <code>true_beta</code> which is
each SNP’s simulated true association value. This data frame is most
useful if we wish to check how well our winner’s curse correction
methods are performing. <code>disc</code> has three columns representing
the summary statistics one would obtain in a discovery GWAS. For each
SNP, this data frame contains its ID number, its estimated effect size
and associated standard error. We note that this function outputs the
set of summary statistics, <code>disc</code>, in a suitable way so that
the winner’s curse correction methods can be directly applied, i.e. in
the form of a <strong>data frame with three columns <code>rsid</code>,
<code>beta</code> and <code>se</code>.</strong></p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1998</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_stats.html">sim_stats</a></span><span class="op">(</span>nsnp<span class="op">=</span><span class="fl">10</span><span class="op">^</span><span class="fl">6</span>,h2<span class="op">=</span><span class="fl">0.4</span>,prop_effect<span class="op">=</span><span class="fl">0.01</span>,nid<span class="op">=</span><span class="fl">50000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## simulated GWAS summary statistics</span></span>
<span><span class="va">summary_stats</span> <span class="op">&lt;-</span> <span class="va">sim_dataset</span><span class="op">$</span><span class="va">disc</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">summary_stats</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rsid         beta          se</span></span>
<span><span class="co">#&gt; 1    1 -0.013919699 0.019222524</span></span>
<span><span class="co">#&gt; 2    2  0.003097741 0.009187098</span></span>
<span><span class="co">#&gt; 3    3 -0.037870279 0.016459187</span></span>
<span><span class="co">#&gt; 4    4 -0.011874324 0.019560506</span></span>
<span><span class="co">#&gt; 5    5  0.024414134 0.006334964</span></span>
<span><span class="co">#&gt; 6    6 -0.001690693 0.007697227</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="method-1-conditional-likelihood">Method 1: Conditional Likelihood<a class="anchor" aria-label="anchor" href="#method-1-conditional-likelihood"></a>
</h3>
<ul>
<li><p>The function <code>conditional_likelihood</code> requires a data
frame, in the form described above, and a value for <code>alpha</code>,
the significance threshold used in the GWAS. Note that all columns of
the data frame must contain numerical values and each row must represent
a unique SNP, identified by <code>rsid</code>.</p></li>
<li><p>As the conditional likelihood methods have been designed to be
only used for SNPs which are deemed significant, the data frame returned
contains only SNPs that have <span class="math inline">\(p\)</span>-values below the inputted genome-wide
significance threshold value, <code>alpha</code>.</p></li>
<li><p><strong>Note:</strong> If the <span class="math inline">\(z\)</span>-statistic of a particular SNP is
greater than 100, then merely the original naive estimate will be
outputted for the second form of the adjusted estimate, namely
<code>beta.cl2</code>, for that SNP.</p></li>
<li><p>If no SNPs are detected as significant, the function returns a
warning message:
<code>WARNING: There are no significant SNPs at this threshold</code>.</p></li>
<li><p>The returned data frame has SNPs ordered based on their <span class="math inline">\(p\)</span>-values from smallest to largest, or
equivalently, in descending order of <span class="math inline">\(\mid z
\mid\)</span> in which <span class="math inline">\(z\)</span> is the
estimated value for <span class="math inline">\(\beta\)</span> divided
by its standard error.</p></li>
<li><p>We implement the function on our toy data set as follows,
choosing a significance threshold of <code>5e-8</code>:</p></li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_CL</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/conditional_likelihood.html">conditional_likelihood</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">summary_stats</span>, alpha <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rsid        beta          se    beta.cl1    beta.cl2    beta.cl3</span></span>
<span><span class="co">#&gt; 1 3965  0.05013020 0.006345754  0.04999615  0.04944871  0.04972243</span></span>
<span><span class="co">#&gt; 2 7815  0.05040019 0.006567997  0.05015587  0.04937802  0.04976694</span></span>
<span><span class="co">#&gt; 3 4998 -0.04856148 0.006458450 -0.04821488 -0.04727195 -0.04774341</span></span>
<span><span class="co">#&gt; 4 7261  0.04649358 0.006342549  0.04597095  0.04478257  0.04537676</span></span>
<span><span class="co">#&gt; 5 6510  0.04611661 0.006499584  0.04522448  0.04357188  0.04439818</span></span>
<span><span class="co">#&gt; 6 9917  0.04489474 0.006333883  0.04401180  0.04238612  0.04319896</span></span></code></pre></div>
<ul>
<li><p>Three columns have been added to the right of the inputted data
frame: each represents a correction method suggested in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2665019/" class="external-link">Ghosh
<em>et al.</em> (2008)</a>. The first, <code>beta.cl1</code>, is
referred to as the conditional MLE while the second,
<code>beta.cl2</code>, is the mean of the normalized conditional
likelihood. The third, <code>beta.cl3</code> is merely the average of
the first two, known as the compromise estimator.</p></li>
<li><p>Ghosh <em>et al.</em> (2008) noted that for instances in which
the true <span class="math inline">\(\beta\)</span> value is close to
zero, <code>beta.cl2</code> often has greater mean squared error than
<code>beta.cl1</code> but for true <span class="math inline">\(\beta\)</span> values further from zero,
<code>beta.cl2</code> performs better. Thus, it was suggested that
<code>beta.cl3</code> be used to combine the strengths of these two
estimators. However, this function, <code>conditional_likelihood</code>,
outputs values for all three estimators in order to allow the user to
make their own decision on which they believe to be the most
appropriate.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="method-2-empirical-bayes">Method 2: Empirical Bayes<a class="anchor" aria-label="anchor" href="#method-2-empirical-bayes"></a>
</h3>
<ul>
<li><p>The function <code>empirical_bayes</code> implements the
empirical Bayes method for correcting for Winner’s Curse, detailed in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4048064/" class="external-link">Ferguson
<em>et al.</em> (2013)</a>, with a slight modification.</p></li>
<li><p>The function has two arguments with the first being
<code>summary_data</code>, the data frame containing <code>rsid</code>,
<code>beta</code> and <code>se</code> columns. Again, all columns of the
data frame must contain numerical values and each row must represent a
unique SNP, identified by <code>rsid</code>. In addition, the function
requires that the data frame must contain at least 50 rows corresponding
to 50 unique SNPs. This requirement is included as the empirical Bayes
method performs poorly when only data of a small number of SNPs is
available. The second argument, <code>method</code>, is a string
parameter which allows the user to choose what modelling approach to
take for the purpose of estimating the log density function. The default
setting is <code>method="AIC"</code>, which is a method very similar to
that described by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4048064/" class="external-link">Ferguson
<em>et al.</em> (2013)</a>. Other options include
<code>method="fix_df"</code>, <code>method="scam"</code>,
<code>method="gam_nb"</code> and <code>method="gam_po"</code>. If
<code>method="fix_df"</code>, the degrees of freedom is set to 7. The
other three options all enforce additional constraints on the shape of
the estimated log density function by i) fitting a generalized additive
model (GAM), with integrated smoothness estimation,
(<code>method="gam_nb"</code> and <code>method="gam_po"</code>) and ii)
applying two shape constrained additive models (SCAMs)
(<code>method="scam"</code>). These additional variations of the
empirical Bayes approach are detailed in the <em>‘Materials and
methods’</em> section of <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>.</p></li>
<li><p><code>empirical_Bayes</code> returns the above described data
frame but with a fourth column, <code>beta_EB</code> in which the
adjusted estimates of this procedure have been added.</p></li>
<li><p>As the empirical Bayes method makes adjustments to all SNPs, not
only those that have been considered significant, this data frame
contains all SNPs.</p></li>
<li><p><strong>Note:</strong> The conditional likelihood methods adjust
each statistic one at a time, and so summary statistics of one
individual SNP can easily be inputted into the
<code>conditional_likelihood</code> function. However, the empirical
Bayes method requires many SNPs as the more information given to the
function, the more accurate it will be at modelling the distribution and
thus, making adjustments. Thus, as mentioned above,
<code>summary_data</code> must contain information related to more than
50 SNPs.</p></li>
<li><p>Below is a demonstration of using <code>empirical_bayes</code>
with our toy data set with the default setting for
<code>method</code>:</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_EB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/empirical_bayes.html">empirical_bayes</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">summary_stats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out_EB</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rsid        beta          se     beta_EB</span></span>
<span><span class="co">#&gt; 1 3965  0.05013020 0.006345754  0.04721273</span></span>
<span><span class="co">#&gt; 2 7815  0.05040019 0.006567997  0.04738054</span></span>
<span><span class="co">#&gt; 3 4998 -0.04856148 0.006458450 -0.04196316</span></span>
<span><span class="co">#&gt; 4 7261  0.04649358 0.006342549  0.04357758</span></span>
<span><span class="co">#&gt; 5 6510  0.04611661 0.006499584  0.04312841</span></span>
<span><span class="co">#&gt; 6 9917  0.04489474 0.006333883  0.04198272</span></span></code></pre></div>
<ul>
<li><p>Unfortunately, the Empirical Bayes estimator originally described
by Ferguson <em>et al.</em> (2013) is known to perform poorly in the
extreme tails of the distribution. Therefore, it is quite possible that
a lack of adjustment for the <span class="math inline">\(\beta\)</span>
values of SNPs with the most extreme <span class="math inline">\(z\)</span>-statistics could be witnessed with this
method. A somewhat ad hoc approach to overcome this issue of combining
the Empirical Bayes and the conditional likelihood methods was suggested
by Ferguson <em>et al.</em> (2013).</p></li>
<li><p>In order to ensure that shrinkage does occur for the <span class="math inline">\(\beta\)</span> values of these extreme SNPs, it
was decided that the following modification would be added to the
<code>empirical_bayes</code> function with <code>method="AIC"</code>.
The basis function of the natural cubic spline has been altered so that
the boundary knots are no longer the most extreme <span class="math inline">\(z\)</span>-values. Instead, the lower boundary
knot is defined as the <span class="math inline">\(10^{\text{th}}\)</span> <span class="math inline">\(z\)</span>-statistic when the <span class="math inline">\(z\)</span>-statistics lie in increasing order
while the upper boundary knot is the <span class="math inline">\(10^{\text{th}}\)</span> <span class="math inline">\(z\)</span>-statistic when the <span class="math inline">\(z\)</span>-statistics have been arranged in
decreasing order. The natural cubic spline is then merely linear beyond
these boundary knots. This assists in fixing the ‘tails’ issue discussed
above. We see that the reduction of the estimated <span class="math inline">\(\beta\)</span> values for all of the top 6 most
significant SNPs has occurred.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="method-3-fdr-inverse-quantile-transformation">Method 3: FDR Inverse Quantile Transformation<a class="anchor" aria-label="anchor" href="#method-3-fdr-inverse-quantile-transformation"></a>
</h3>
<ul>
<li><p>The function <code>FDR_IQT</code> implements the winner’s curse
adjustment method detailed in <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5013908/" class="external-link">Bigdeli
<em>et al.</em> (2016)</a>.</p></li>
<li><p>Similar to the approaches above, the function requires a data
frame with three columns <code>rsid</code>, <code>beta</code> and
<code>se</code>, all columns of the data frame must contain numerical
values and each row must represent a unique SNP, identified by
<code>rsid</code>.</p></li>
<li><p>It also has an argument <code>min_pval</code>, for which the
default setting is <code>1e-300</code>. This is included in order to
avoid zero <span class="math inline">\(p\)</span>-values which can prove
problematic for the function when evaluating <code><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm()</a></code>.
Furthermore, due to the nature of winner’s curse, it is in fact rare
that observations with <span class="math inline">\(\mid z \mid &gt;
37\)</span> are biased.</p></li>
<li><p>The function outputs a data frame similar to that inputted with
an additional column containing the adjusted estimate,
<code>beta_FIQT</code>, and again, the SNPs have been reordered
according to their individual <span class="math inline">\(z\)</span>-statistics.</p></li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_FIQT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FDR_IQT.html">FDR_IQT</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">summary_stats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out_FIQT</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rsid        beta          se   beta_FIQT</span></span>
<span><span class="co">#&gt; 1 3965  0.05013020 0.006345754  0.03771470</span></span>
<span><span class="co">#&gt; 2 7815  0.05040019 0.006567997  0.03783823</span></span>
<span><span class="co">#&gt; 3 4998 -0.04856148 0.006458450 -0.03633847</span></span>
<span><span class="co">#&gt; 4 7261  0.04649358 0.006342549  0.03442008</span></span>
<span><span class="co">#&gt; 5 6510  0.04611661 0.006499584  0.03364166</span></span>
<span><span class="co">#&gt; 6 9917  0.04489474 0.006333883  0.03278400</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="method-4-bootstrap">Method 4: Bootstrap<a class="anchor" aria-label="anchor" href="#method-4-bootstrap"></a>
</h3>
<ul>
<li><p>Inspired by the bootstrap resampling method first detailed in <a href="https://europepmc.org/article/med/21538984" class="external-link">Faye <em>et al.</em>
(2011)</a>, the function <code>BR_ss</code> implements a similar
approach which can be easily applied to published sets of GWAS summary
statistics without the requirement of access to original
individual-level data. The major advantage of our alternative suggested
method is the considerable improvement in computational efficiency, with
a reasonably low level of conceptual difficulty being maintained. A
comprehensive description of this new bootstrap method can be found in
the <em>‘Materials and methods’</em> section of <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>.</p></li>
<li><p><code>BR_ss</code> takes a data frame <code>summary_data</code>
with three columns: <code>rsid</code>, <code>beta</code>,
<code>se</code>, and returns the inputted data frame along with the
adjusted estimate for <span class="math inline">\(\beta\)</span>,
<code>beta_BR_ss</code>. Note that all columns of the data frame,
<code>summary data</code> must contain numerical values, each row must
represent a unique SNP, identified by <code>rsid</code> and the function
requires that there must be at least 5 unique SNPs so that the smoothing
spline can be implemented successfully. The function also has a logical
parameter <code>seed_opt</code> which can be set to <code>TRUE</code> by
the user to permit reproducibility. Its default setting is
<code>seed_opt=FALSE</code>. As this method involves bootstrap
implementation, the use of a seed can allow the user to obtain the exact
same results for <code>beta_BR_ss</code> when running the function many
times on the same dataset. The parameter <code>seed</code> provides the
user with an option to set their desired value for the seed when using
<code>seed_opt=TRUE</code>. The default for <code>seed</code> is the
arbitrary value of <code>1998</code>.</p></li>
</ul>
<p><span class="math inline">\(~\)</span></p>
<p><strong>Brief method description:</strong></p>
<ul>
<li>
<p>This procedure begins with arranging all <span class="math inline">\(N\)</span> SNPs according to their original <span class="math inline">\(z\)</span>-statistics, <span class="math inline">\(z_i =
\frac{\hat\beta_i}{\hat{\text{se}(\hat\beta_i)}}\)</span>, in descending
order. A randomized estimate of the extent of ranking bias for the <span class="math inline">\(k^{th}\)</span> largest <span class="math inline">\(z\)</span>-statistic is calculated by means of the
parametric bootstrap as follows:</p>
<ol style="list-style-type: decimal">
<li><p>A value <span class="math inline">\(\hat\beta_i^{\text{b}}\)</span> is simulated for
SNP <span class="math inline">\(i, i=1,...,N\)</span>, independently,
from a Gaussian distribution with mean <span class="math inline">\(\hat\beta_i\)</span> and standard deviation <span class="math inline">\(\hat{\text{se}(\hat\beta_i)}\)</span>, i.e. <span class="math display">\[\hat\beta_i^{\text{b}} \sim N(\hat\beta_i,
\hat{\text{se}(\hat\beta_i)}).\]</span></p></li>
<li><p>Upon obtaining <span class="math inline">\(\hat\beta_i^{\text{b}}\)</span> for <span class="math inline">\(i, i=1,...,N\)</span>, the SNPs are subjected to
an ordering in which they are arranged based on their <span class="math inline">\(z_i^{\text{b}}\)</span>-statistics, in descending
order. The <span class="math inline">\(z_i^{\text{b}}\)</span>-statistic
of SNP <span class="math inline">\(i\)</span> is defined as <span class="math display">\[z_i^{\text{b}} =
\frac{\hat\beta_i^{\text{b}}}{\hat{\text{se}(\hat\beta_i)}}.\]</span>Due
to this ordering, each SNP is assigned a rank <span class="math inline">\(A(k)\)</span>. Equivalently, we define <span class="math inline">\(A(k)\)</span> as the index corresponding to the
<span class="math inline">\(k^{th}\)</span> largest entry in the vector:
<span class="math inline">\([z_1^{\text{b}}, ..., z_N^{\text{b}}] =
[\frac{\hat\beta_1^{\text{b}}}{\hat{\text{se}(\hat\beta_1)}}, ...,
\frac{\hat\beta_N^{\text{b}}}{\hat{\text{se}(\hat\beta_N)}}]\)</span>.</p></li>
<li><p>Then, the estimated bias of SNP <span class="math inline">\(k\)</span>, the SNP with the <span class="math inline">\(k^{th}\)</span> largest original <span class="math inline">\(z\)</span>-statistic, takes the following form:
<span class="math display">\[\text{bias}_k =
\frac{\hat\beta_{A(k)}^{\text{b}}-
\hat\beta_{A(k)}^{\text{oob}}}{\hat{\text{se}(\hat\beta_{A(k)})}} =
\frac{\hat\beta_{A(k)}^{\text{b}}-
\hat\beta_{A(k)}}{\hat{\text{se}(\hat\beta_{A(k)})}},\]</span>in which
<span class="math inline">\(\hat\beta_{A(k)}^{\text{b}}\)</span> is the
bootstrap value of the SNP ranked in position <span class="math inline">\(k\)</span> in the ordering of <span class="math inline">\(z_i^{\text{b}}\)</span>-statistics, <span class="math inline">\(\hat\beta_{A(k)}^{\text{oob}} =
\hat\beta_{A(k)}\)</span> is that same SNP’s original <span class="math inline">\(\beta\)</span> estimate and <span class="math inline">\(\hat{\text{se}(\hat\beta_{A(k)})}\)</span> its
standard error.</p></li>
</ol>
</li>
<li><p>In the next step of the process, a cubic smoothing spline is
fitted to the data in which the <span class="math inline">\(z\)</span>-statistics are considered as the inputs
and <span class="math inline">\(\text{bias}_k\)</span>, their
corresponding outputs. The predicted values from this model fitting
provides new estimates for the bias correction, <span class="math inline">\(\text{bias}_k^*\)</span> for each SNP. This
additional stage in which <span class="math inline">\(\text{bias}_k^*\)</span> is obtained eliminates
the need for several bootstrap iterations for each SNP in order to
ensure competitive performance of the method. Thus, this results in an
approach that is faster as only one bootstrap iteration is required per
SNP, while using <span class="math inline">\(\text{bias}_k^*\)</span>
instead of <span class="math inline">\(\text{bias}_k\)</span> ensures
that the method has greater accuracy.</p></li>
<li><p>Finally, the new estimate for the true effect size of SNP k, the
SNP with the <span class="math inline">\(k^{\text{th}}\)</span> largest
original <span class="math inline">\(z\)</span>-statistic, is defined as
<span class="math inline">\(\hat\beta_k^* = \hat\beta_k -
\hat{\text{se}(\hat\beta_k)} \cdot \text{bias}_k^*\)</span>.</p></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_BR_ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BR_ss.html">BR_ss</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">summary_stats</span>, seed_opt<span class="op">=</span><span class="cn">TRUE</span>, seed<span class="op">=</span><span class="fl">2020</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out_BR_ss</span><span class="op">)</span></span>
<span><span class="co">#&gt;   rsid        beta          se  beta_BR_ss</span></span>
<span><span class="co">#&gt; 1 3965  0.05013020 0.006345754  0.04173996</span></span>
<span><span class="co">#&gt; 2 7815  0.05040019 0.006567997  0.04114148</span></span>
<span><span class="co">#&gt; 3 4998 -0.04856148 0.006458450 -0.03687195</span></span>
<span><span class="co">#&gt; 4 7261  0.04649358 0.006342549  0.03668250</span></span>
<span><span class="co">#&gt; 5 6510  0.04611661 0.006499584  0.03544370</span></span>
<span><span class="co">#&gt; 6 9917  0.04489474 0.006333883  0.03447524</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparing-results">Comparing Results<a class="anchor" aria-label="anchor" href="#comparing-results"></a>
</h3>
<p>As we have simulated our data set, we have the ability to quantify
the amount of bias present in the original naive <span class="math inline">\(\beta\)</span> estimates and also, in our adjusted
estimates for each method. Thus, we can briefly examine how each
correction method performed with respect to our simulated data set.
Similar to <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>, assessment can take place using measures
such as the estimated root mean square error of significant SNPs
(<code>rmse</code>) and the estimated average bias over all significant
SNPs (<code>bias</code>). These two evaluation metrics can be be
mathematically defined as follows, in which there are <span class="math inline">\(N_{\text{sig}}\)</span> significant SNPs:</p>
<ul>
<li>
<code>rmse</code> <span class="math inline">\(=
\sqrt{\frac{1}{N_{\text{sig}}}\sum_{i=1}^{N_{\text{sig}}}
(\hat\beta_{\text{adj},i} - \beta_i)^2}\)</span>
</li>
<li>
<code>bias</code> <span class="math inline">\(=
\frac{1}{N_{\text{sig}}} \sum_{i=1}^{N_{\text{sig}}} (
\hat\beta_{\text{adj},i}- \beta_i)\)</span>
</li>
</ul>
<p><span style="color: blue;"><span class="math inline">\(\star\)</span></span> <strong>Note:</strong> This
average bias will be computed separately for significant SNPs with
positive and negative true effects to avoid the bias in either direction
simply cancelling each other out.</p>
<p><span style="color: blue;"><span class="math inline">\(\star\)</span></span> <strong>Note:</strong> We
compare our methods based on their performance with respect to SNPs that
pass a certain threshold as these are the SNPs that have association
estimates affected by winner’s curse and most often, it is the effect
sizes of these SNPs that researchers are most interested in.</p>
<p>We consider a significance threshold of <span class="math inline">\(\alpha = 5 \times 10^{-8}\)</span> and subset the
outputted data frames as required.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Simulated true effect sizes:</span></span>
<span><span class="va">true_beta</span> <span class="op">&lt;-</span> <span class="va">sim_dataset</span><span class="op">$</span><span class="va">true</span><span class="op">$</span><span class="va">true_beta</span> </span>
<span></span>
<span><span class="co">## Subset required data sets:</span></span>
<span><span class="va">out_EB_sig</span> <span class="op">&lt;-</span> <span class="va">out_EB</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_EB</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_EB</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span>
<span><span class="va">out_FIQT_sig</span> <span class="op">&lt;-</span> <span class="va">out_FIQT</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_FIQT</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_FIQT</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span>
<span><span class="va">out_BR_ss_sig</span> <span class="op">&lt;-</span> <span class="va">out_BR_ss</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_BR_ss</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_BR_ss</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span></code></pre></div>
<p>The estimated root mean square error of significant SNPs for each
method is computed below. It can be seen that all methods demonstrate
improvements over merely using the naive unadjusted association
estimates as the <code>rmse</code> values for each method are lower than
that of <code>naive</code>. The bootstrap method provides the lowest
value for <code>rmse</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rmse</span></span>
<span><span class="va">rmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>naive <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, CL1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl1</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, CL2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl2</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, CL3<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl3</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,EB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_EB_sig</span><span class="op">$</span><span class="va">beta_EB</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_EB_sig</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, FIQT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_FIQT_sig</span><span class="op">$</span><span class="va">beta_FIQT</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_FIQT_sig</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, boot<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">out_BR_ss_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">out_BR_ss_sig</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rmse</span></span>
<span><span class="co">#&gt;        naive        CL1        CL2        CL3         EB        FIQT</span></span>
<span><span class="co">#&gt; 1 0.01486281 0.01277932 0.00875216 0.01049765 0.01080075 0.007897031</span></span>
<span><span class="co">#&gt;          boot</span></span>
<span><span class="co">#&gt; 1 0.007715133</span></span></code></pre></div>
<p>The next metric, the average bias over all significant SNPs with
positive association estimates, is included below. Here, we wish to
obtain values closer to zero than the <code>naive</code> approach and it
can be seen that all methods achieve this.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## bias positive</span></span>
<span><span class="va">pos_sig</span> <span class="op">&lt;-</span> <span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">pos_sigA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">pos_sig</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bias_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>naive <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>, CL1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl1</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>, CL2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl2</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>, CL3<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl3</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>,EB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_EB_sig</span><span class="op">$</span><span class="va">beta_EB</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>, FIQT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_FIQT_sig</span><span class="op">$</span><span class="va">beta_FIQT</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span>, boot<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_BR_ss_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span><span class="op">[</span><span class="va">pos_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">pos_sig</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bias_up</span></span>
<span><span class="co">#&gt;       naive         CL1        CL2         CL3          EB         FIQT</span></span>
<span><span class="co">#&gt; 1 0.0126302 0.002708404 0.00279926 0.002753832 0.008984969 -0.001100354</span></span>
<span><span class="co">#&gt;          boot</span></span>
<span><span class="co">#&gt; 1 0.000130083</span></span></code></pre></div>
<p>In a similar manner, the average bias over all significant SNPs with
negative association estimates, is computed. Again, we see that the
application of all methods result in, on average, less biased
association estimates for these SNPs, as desired.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## bias negative</span></span>
<span><span class="va">neg_sig</span> <span class="op">&lt;-</span> <span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">neg_sigA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">neg_sig</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bias_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>naive <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>, CL1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl1</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>, CL2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl2</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>, CL3<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl3</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>,EB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_EB_sig</span><span class="op">$</span><span class="va">beta_EB</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>, FIQT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_FIQT_sig</span><span class="op">$</span><span class="va">beta_FIQT</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span>, boot<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_BR_ss_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span><span class="op">[</span><span class="va">neg_sigA</span><span class="op">]</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">neg_sig</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bias_down</span></span>
<span><span class="co">#&gt;         naive         CL1          CL2         CL3           EB       FIQT</span></span>
<span><span class="co">#&gt; 1 -0.01172448 0.001909202 0.0003890267 0.001149115 -0.004655593 0.00188084</span></span>
<span><span class="co">#&gt;          boot</span></span>
<span><span class="co">#&gt; 1 0.001905609</span></span></code></pre></div>
<p><span style="color: blue;"><span class="math inline">\(\star\)</span></span> <strong>Note:</strong> In
the above results, it is seen that the empirical Bayes method doesn’t
seem to be performing as well as the other methods, i.e. the reductions
in <code>rmse</code> and <code>bias</code> are small in comparison. The
reason for this is perhaps that it is the original form of the empirical
Bayes method that has been used here. It is possible that using another
variation of the method, for example setting
<code>method="gam_nb"</code> or <code>method="scam"</code>, would result
in improvements, as shown in <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>.</p>
<p>To complement the above, we provide an illustration of boxplots
obtained for the root mean square error of significant SNPs
(<strong>RMSE of sig. SNPs at <span class="math inline">\(5 \times
10^{-8}\)</span></strong>), plotted for each winner’s curse correction
method (<strong>Method</strong>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">8</span>,<span class="st">"Dark2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rsid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, se <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">se</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, adj_beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span>,<span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl1</span>,<span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl2</span>,<span class="va">out_CL</span><span class="op">$</span><span class="va">beta.cl3</span>,<span class="va">out_EB_sig</span><span class="op">$</span><span class="va">beta_EB</span>,<span class="va">out_FIQT_sig</span><span class="op">$</span><span class="va">beta_FIQT</span>,<span class="va">out_BR_ss_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span><span class="op">)</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL1"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL2"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL3"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"EB"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FIQT"</span>,<span class="fl">34</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"boot"</span>,<span class="fl">34</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_results</span><span class="op">$</span><span class="va">rmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="op">(</span><span class="va">all_results</span><span class="op">$</span><span class="va">adj_beta</span> <span class="op">-</span> <span class="va">true_beta</span><span class="op">[</span><span class="va">all_results</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">all_results</span><span class="op">$</span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">all_results</span><span class="op">$</span><span class="va">method</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="st">"CL1"</span>, <span class="st">"CL2"</span>, <span class="st">"CL3"</span>, <span class="st">"EB"</span>, <span class="st">"boot"</span>, <span class="st">"FIQT"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">all_results</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">method</span>,y<span class="op">=</span><span class="va">rmse</span>,fill<span class="op">=</span><span class="va">method</span>, color<span class="op">=</span><span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.7</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">method</span>, color<span class="op">=</span><span class="va">method</span>, alpha<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span>,<span class="va">col</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Method"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"RMSE of sig. SNPs at  "</span>, <span class="fl">5</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept<span class="op">=</span><span class="fl">0</span>, colour<span class="op">=</span><span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">90</span>, vjust<span class="op">=</span><span class="fl">0.5</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,legend.position <span class="op">=</span> <span class="st">"none"</span>, strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face<span class="op">=</span><span class="st">"italic"</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p><img src="winners_curse_methods_files/figure-html/unnamed-chunk-11-1.png" width="624"></p>
<p>In addition, we can gain a visual insight into the adjustments made
by these functions by constructing a <strong>density</strong> plot with
the adjusted <em>absolute</em> values along with the naive estimates and
the true <em>absolute</em> <span class="math inline">\(\beta\)</span>
values of significant SNPs, as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rsid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">)</span>, beta<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">beta</span><span class="op">)</span>, se<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>,adj_beta<span class="op">=</span><span class="va">true_beta</span><span class="op">[</span><span class="va">out_CL</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span>,method<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"true"</span>,<span class="fl">34</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_resultsA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">all_results</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,<span class="va">true</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">all_resultsA</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">adj_beta</span><span class="op">)</span>,colour<span class="op">=</span><span class="va">method</span>,fill<span class="op">=</span><span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.05</span>,size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>,<span class="fl">0.08</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"|"</span>, <span class="va">beta</span>, <span class="st">"| "</span> , <span class="st">"of sig. SNPs at  "</span>, <span class="fl">5</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Method"</span>,colour<span class="op">=</span><span class="st">"Method"</span><span class="op">)</span></span></code></pre></div>
<p><img src="winners_curse_methods_files/figure-html/unnamed-chunk-12-1.png" width="624"></p>
<p><span class="math inline">\(~\)</span> <span class="math inline">\(~\)</span></p>
</div>
<div class="section level3">
<h3 id="application-of-methods-to-a-real-data-set">Application of methods to a real data set<a class="anchor" aria-label="anchor" href="#application-of-methods-to-a-real-data-set"></a>
</h3>
<p>In this final section, we will provide a short demonstration of
applying these winner’s curse correction methods to a real data set.</p>
<p>Briefly, the data set we will use is a subset of the first BMI data
set used in <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a>, which was obtained from the UK Biobank. In
order to obtain a smaller data set suitable for demonstration purposes
in this vignette, LD clumping was performed on the data set used in the
aforementioned paper using the <code>ld_clump()</code> function from the
<a href="https://mrcieu.github.io/ieugwasr/" class="external-link">ieugwasr R package</a>. The
LD (<em>r</em>-squared) threshold for clumping was set to 0.01 while the
physical distance threshold used was 100 kb. This procedure provided a
data set of 272,604 SNPs with corresponding association estimates
(<code>beta</code>) and standard errors (<code>se</code>) as well as
association estimates obtained in an independent <em>replication</em>
GWAS of a similar size (<code>beta_rep</code>). The first six rows of
this data set are outputted below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BMI_UKBB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/amandaforde/winnerscurse_realdata/main/data/summary_stats_BMI.txt"</span>,quote <span class="op">=</span> <span class="st">""</span>, row.names <span class="op">=</span> <span class="cn">NULL</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="st">"pos"</span>,<span class="st">"rsid"</span>,<span class="st">"beta"</span>,<span class="st">"se"</span>,<span class="st">"beta_rep"</span><span class="op">)</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">BMI_UKBB</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr    pos          rsid        beta        se    beta_rep</span></span>
<span><span class="co">#&gt; 1   1 801536  "rs79373928"  0.00691695 0.0672237 -0.00845073</span></span>
<span><span class="co">#&gt; 2   1 809876  "rs57181708"  0.08085120 0.0271055  0.05181780</span></span>
<span><span class="co">#&gt; 3   1 815656 "rs557970564"  0.16602900 0.0887715 -0.13927300</span></span>
<span><span class="co">#&gt; 4   1 834430 "rs533316052" -0.06663160 0.0685312 -0.05314070</span></span>
<span><span class="co">#&gt; 5   1 837657 "rs149737509" -0.00116905 0.0623581 -0.06175060</span></span>
<span><span class="co">#&gt; 6   1 852758   "rs4970462" -0.02576140 0.0200347 -0.02956890</span></span></code></pre></div>
<p>We apply the correction methods to this data set as follows and
obtain adjusted association estimates for significant SNPs at the
genome-wide significance threshold of <span class="math inline">\(5
\times 10^{-8}\)</span>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BMI_data</span> <span class="op">&lt;-</span> <span class="va">BMI_UKBB</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="va">out_CL_BMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/conditional_likelihood.html">conditional_likelihood</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">BMI_data</span>, alpha <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span></span>
<span><span class="va">out_EB_BMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/empirical_bayes.html">empirical_bayes</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">BMI_data</span><span class="op">)</span></span>
<span><span class="va">out_FIQT_BMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FDR_IQT.html">FDR_IQT</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">BMI_data</span><span class="op">)</span></span>
<span><span class="va">out_boot_BMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BR_ss.html">BR_ss</a></span><span class="op">(</span>summary_data <span class="op">=</span> <span class="va">BMI_data</span>, seed<span class="op">=</span><span class="fl">2020</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Subset required data sets:</span></span>
<span><span class="va">out_EB_BMI_sig</span> <span class="op">&lt;-</span> <span class="va">out_EB_BMI</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_EB_BMI</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_EB_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span>
<span><span class="va">out_FIQT_BMI_sig</span> <span class="op">&lt;-</span> <span class="va">out_FIQT_BMI</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_FIQT_BMI</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_FIQT_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span>
<span><span class="va">out_boot_BMI_sig</span> <span class="op">&lt;-</span> <span class="va">out_boot_BMI</span><span class="op">[</span><span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out_boot_BMI</span><span class="op">$</span><span class="va">beta</span><span class="op">/</span><span class="va">out_boot_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>, lower.tail<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5e-8</span>,<span class="op">]</span></span></code></pre></div>
<p><span style="color: blue;"><span class="math inline">\(\star\)</span></span> <strong>Note:</strong> In
the case of real data, the <strong><em>true</em></strong> association
values are unknown. Therefore, in order to evaluate the performance of
methods, we will make use of <code>beta_rep</code>, the association
estimates obtained in an independent <em>replication</em> GWAS of a
similar size. These association estimates are considered to be
<strong><em>unbiased</em></strong>, and thus can be used to obtain an
expression for the estimated mean square error (MSE) of significant
SNPs. This expression is given in <em>Eq [14]</em> of <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1010546" class="external-link">Forde
<em>et al.</em> (2023)</a> and can be computed for each method as shown
below. It can be seen that each method provides an estimated MSE value
lower than that of the naive approach of using the unadjusted
association estimates from the discovery GWAS.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## mse</span></span>
<span><span class="va">beta_rep</span> <span class="op">&lt;-</span> <span class="va">BMI_UKBB</span><span class="op">$</span><span class="va">beta_rep</span><span class="op">[</span><span class="va">BMI_UKBB</span><span class="op">$</span><span class="va">rsid</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">rsid</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>naive <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, CL1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl1</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, CL2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl2</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, CL3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl3</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, EB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_EB_BMI_sig</span><span class="op">$</span><span class="va">beta_EB</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_EB_BMI_sig</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, boot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_boot_BMI_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_boot_BMI_sig</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>, FIQT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">beta_rep</span> <span class="op">-</span> <span class="va">out_FIQT_BMI_sig</span><span class="op">$</span><span class="va">beta_FIQT</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">out_FIQT_BMI_sig</span><span class="op">$</span><span class="va">se</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mse</span></span>
<span><span class="co">#&gt;        naive        CL1        CL2        CL3         EB       boot       FIQT</span></span>
<span><span class="co">#&gt; 1 0.03608036 0.02946674 0.02870324 0.02897353 0.03029724 0.02874184 0.02859518</span></span></code></pre></div>
<p>For illustration purposes, we again simply produce a
<strong>density</strong> plot with the adjusted <em>absolute</em> values
along with the naive estimates and the true <em>absolute</em> <span class="math inline">\(\beta\)</span> values of significant SNPs.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rsid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">rsid</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, se <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span>, adj_beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta</span>,<span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl1</span>,<span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl2</span>,<span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta.cl3</span>,<span class="va">out_EB_BMI_sig</span><span class="op">$</span><span class="va">beta_EB</span>,<span class="va">out_FIQT_BMI_sig</span><span class="op">$</span><span class="va">beta_FIQT</span>,<span class="va">out_boot_BMI_sig</span><span class="op">$</span><span class="va">beta_BR_ss</span><span class="op">)</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL1"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL2"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"CL3"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"EB"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FIQT"</span>,<span class="fl">214</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"boot"</span>,<span class="fl">214</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_results</span><span class="op">$</span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">all_results</span><span class="op">$</span><span class="va">method</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="st">"CL1"</span>, <span class="st">"CL2"</span>, <span class="st">"CL3"</span>, <span class="st">"EB"</span>, <span class="st">"boot"</span>, <span class="st">"FIQT"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rsid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">rsid</span><span class="op">)</span>, beta<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">beta</span><span class="op">)</span>, se<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">out_CL_BMI</span><span class="op">$</span><span class="va">se</span><span class="op">)</span>,adj_beta<span class="op">=</span><span class="va">beta_rep</span>,method<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"rep"</span>,<span class="fl">214</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_resultsA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">all_results</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,<span class="va">rep</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">all_resultsA</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">adj_beta</span><span class="op">)</span>,colour<span class="op">=</span><span class="va">method</span>,fill<span class="op">=</span><span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.05</span>,size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.1</span>,<span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"|"</span>, <span class="va">beta</span>, <span class="st">"| "</span> , <span class="st">"of sig. SNPs at  "</span>, <span class="fl">5</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Method"</span>,colour<span class="op">=</span><span class="st">"Method"</span><span class="op">)</span></span></code></pre></div>
<p><img src="winners_curse_methods_files/figure-html/unnamed-chunk-16-1.png" width="624"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Amanda Forde.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
