<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>FDR IQT method for use with discovery GWAS — FDR_IQT • winnerscurse</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="FDR IQT method for use with discovery GWAS — FDR_IQT" />
<meta property="og:description" content="FDR_IQT is a function which uses summary statistics to reduce Winner's
Curse bias in SNP-trait association estimates, obtained from a discovery GWAS.
The function implements the FDR Inverse Quantile Transformation method
described in
Bigdeli et
al. (2016), which was established for this purpose." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">winnerscurse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/discovery_replication.html">Using discovery and replication summary statistics for adjusting association estimates</a>
    </li>
    <li>
      <a href="../articles/standard_errors_confidence_intervals.html">Standard errors of adjusted estimates and confidence intervals</a>
    </li>
    <li>
      <a href="../articles/winners_curse_methods.html">Obtaining adjusted `beta` estimates</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>FDR IQT method for use with discovery GWAS</h1>
    
    <div class="hidden name"><code>FDR_IQT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>FDR_IQT</code> is a function which uses summary statistics to reduce Winner's
Curse bias in SNP-trait association estimates, obtained from a discovery GWAS.
The function implements the FDR Inverse Quantile Transformation method
described in
<a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5013908/'>Bigdeli <em>et
al.</em> (2016)</a>, which was established for this purpose.</p>
    </div>

    <pre class="usage"><span class='fu'>FDR_IQT</span><span class='op'>(</span><span class='va'>summary_data</span>, min_pval <span class='op'>=</span> <span class='fl'>1e-15</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>summary_data</th>
      <td><p>A data frame containing summary statistics from the
discovery GWAS. It must have three columns with column names <code>rsid</code>,
<code>beta</code> and <code>se</code>, respectively, and all columns must contain
numerical values.</p></td>
    </tr>
    <tr>
      <th>min_pval</th>
      <td><p>A numerical value whose purpose is to avoid zero
\(p\)-values as this introduces issues when <code><a href='https://rdrr.io/r/stats/Normal.html'>qnorm()</a></code> is applied. Any
SNP for which its computed \(p\)-value is found to be less than
<code>min_pval</code> is merely re-assigned <code>min_pval</code> as its \(p\)-value
and the method proceeds. By definition, the method makes no adjustment to
the association estimate of a SNP for which this has occurred with the
presumption that in general, estimates of SNPs with \(z &gt; 8\) are not
biased. The default value is <code>min_pval = 1e-15</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with the inputted summary data occupying the first three
 columns. The new adjusted association estimates for each SNP are returned in
 the fourth column, namely <code>beta_FIQT</code>. The SNPs are contained in this
 data frame according to their significance, with the most significant SNP,
 i.e. the SNP with the largest absolute \(z\)-statistic, now located in the
 first row of the data frame.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Bigdeli, T. B., Lee, D., Webb, B. T., Riley, B. P., Vladimirov, V.
 I., Fanous, A. H., Kendler, K. S., &amp; Bacanu, S. A. (2016). A simple yet
 accurate correction for winner's curse can predict signals discovered in
 much larger genome scans. <em>Bioinformatics (Oxford, England)</em>,
 <strong>32(17)</strong>, 2598\(-\)2603.
 <a href='https://doi.org/10.1093/bioinformatics/btw303'>https://doi.org/10.1093/bioinformatics/btw303</a></p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amanda Forde.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


