<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>UMVCUE method for use with discovery and replication GWASs — UMVCUE • winnerscurse</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="UMVCUE method for use with discovery and replication GWASs — UMVCUE"><meta property="og:description" content="UMVCUE is a function which aims to produce less biased SNP-trait
association estimates for SNPs deemed significant in the discovery GWAS, using
summary statistics from both discovery and replication GWASs. The function
implements the method described in
Bowden and
Dudbridge (2009), which was established for this purpose."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">winnerscurse</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/discovery_replication.html">Methods for use with discovery and replication GWASs</a>
    </li>
    <li>
      <a href="../articles/standard_errors_confidence_intervals.html">Standard errors and confidence intervals of adjusted estimates</a>
    </li>
    <li>
      <a href="../articles/winners_curse_methods.html">Methods for use with discovery GWAS</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>UMVCUE method for use with discovery and replication GWASs</h1>
    
    <div class="hidden name"><code>UMVCUE.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>UMVCUE</code> is a function which aims to produce less biased SNP-trait
association estimates for SNPs deemed significant in the discovery GWAS, using
summary statistics from both discovery and replication GWASs. The function
implements the method described in
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2726957/" class="external-link">Bowden and
Dudbridge (2009)</a>, which was established for this purpose.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">UMVCUE</span><span class="op">(</span><span class="va">summary_disc</span>, <span class="va">summary_rep</span>, alpha <span class="op">=</span> <span class="fl">5e-08</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>summary_disc</dt>
<dd><p>A data frame containing summary statistics from the
<em>discovery</em> GWAS. It must have three columns with column names
<code>rsid</code>, <code>beta</code> and <code>se</code>, respectively, and columns
<code>beta</code> and <code>se</code> must contain numerical values. Each row must
correspond to a unique SNP, identified by <code>rsid</code>.</p></dd>
<dt>summary_rep</dt>
<dd><p>A data frame containing summary statistics from the
<em>replication</em> GWAS. It must have three columns with column names
<code>rsid</code>, <code>beta</code> and <code>se</code>, respectively, and all columns must
contain numerical values. Each row must correspond to a unique SNP,
identified by the numerical value <code>rsid</code>. SNPs must be ordered in the
exact same manner as those in <code>summary_disc</code>, i.e.
<code>summary_rep$rsid</code> must be equivalent to <code>summary_disc$rsid</code>.</p></dd>
<dt>alpha</dt>
<dd><p>A numerical value which specifies the desired genome-wide
significance threshold for the discovery GWAS. The default is given as
<code>5e-8</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame with summary statistics and adjusted association estimate
 of only those SNPs which have been deemed significant in the discovery GWAS
 according to the specified threshold, <code>alpha</code>, i.e. SNPs with
\(p\)-values less than <code>alpha</code>. The inputted summary data occupies
 the first five columns, in which the columns <code>beta_disc</code> and
<code>se_disc</code> contain the statistics from the discovery GWAS and columns
<code>beta_rep</code> and <code>se_rep</code> hold the replication GWAS statistics. The
 new adjusted association estimate for each SNP, as defined in the
 aforementioned paper, is contained in the final column, namely
<code>beta_UMVCUE</code>. The SNPs are contained in this data frame according to
 their significance, with the most significant SNP, i.e. the SNP with the
 largest absolute \(z\)-statistic, now located in the first row of the data
 frame. If no SNPs are detected as significant in the discovery GWAS,
<code>UMVCUE</code> merely returns a data frame which combines the two inputted
 data sets.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Bowden, J., &amp; Dudbridge, F. (2009). Unbiased estimation of odds
 ratios: combining genomewide association scans with replication studies.
 <em>Genetic epidemiology</em>, <strong>33(5)</strong>, 406\(-\)418.
 <a href="https://doi.org/10.1002/gepi.20394" class="external-link">https://doi.org/10.1002/gepi.20394</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="https://amandaforde.github.io/winnerscurse/articles/discovery_replication.html" class="external-link">https://amandaforde.github.io/winnerscurse/articles/discovery_replication.html</a>
for illustration of the use of <code>UMVCUE</code> with toy data sets and further
information regarding computation of the adjusted SNP-trait association
estimates for significant SNPs.</p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Amanda Forde.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

